src/Compilers/CSharp/Portable/Binder/Binder_Await.cs

```diff
@@ -286,7 +302,130 @@ private bool GetAwaitableExpressionInfo(
+                {
+                    awaitCall = SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__AwaitValueTaskT_T;
+                    maybeNestedType = ((NamedTypeSymbol)expression.Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                }
+                else
+                {
+                    runtimeAwaitHelper = null;
+                    return false;
+                }
+
+                runtimeAwaitHelper = (MethodSymbol)GetSpecialTypeMember(awaitCall, diagnostics, expression.Syntax);
+
+                if (runtimeAwaitHelper is null)
+                {
+                    return false;
+                }
+
+                if (maybeNestedType is { } nestedType)
+                {
+                    Debug.Assert(runtimeAwaitHelper.TypeParameters.Length == 1);
+                    runtimeAwaitHelper = runtimeAwaitHelper.Construct([nestedType]);
+                    checkMethodGenericConstraints(runtimeAwaitHelper, diagnostics, expression.Syntax.Location);
+                }
+#if DEBUG
+                else
+                {
+                    Debug.Assert(runtimeAwaitHelper.TypeParameters.Length == 0);
+                }
+#endif
+
+                return true;
+            }
+
+            bool getRuntimeAwaitAwaiter(TypeSymbol awaiterType, out MethodSymbol? runtimeAwaitAwaiterMethod)
+            {
+                // Use site info is discarded because we don't actually do this conversion, we just need to know which generic
+                // method to call.
+                var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+                var useUnsafeAwait = Compilation.Conversions.ClassifyImplicitConversionFromType(
+                    awaiterType,
+                    Compilation.GetSpecialType(InternalSpecialType.System_Runtime_CompilerServices_ICriticalNotifyCompletion),
+                    ref discardedUseSiteInfo).IsImplicit;
+
+                var awaitMethod = (MethodSymbol?)GetSpecialTypeMember(
+                    useUnsafeAwait
+                        ? SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__UnsafeAwaitAwaiter_TAwaiter
+                        : SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__AwaitAwaiter_TAwaiter,
+                    diagnostics,
+                    expression.Syntax);
+
+                if (awaitMethod is null)
+                {
+                    runtimeAwaitAwaiterMethod = null;
+                    return false;
+                }
+
+                Debug.Assert(awaitMethod is { Arity: 1 });
+
+                runtimeAwaitAwaiterMethod = awaitMethod.Construct(awaiterType);
+                checkMethodGenericConstraints(runtimeAwaitAwaiterMethod, diagnostics, expression.Syntax.Location);
+
+                return true;
+            }
+
+            void checkMethodGenericConstraints(MethodSymbol method, BindingDiagnosticBag diagnostics, Location location)
```

Comment:
Would this helper from `ConstraintHelper` do the job? `public static bool CheckConstraints(this MethodSymbol method, in CheckConstraintsArgs args)`

Reply:
I looked for such a thing and missed, thanks for pointing it out!